cudax/include/cuda/experimental/__graph/graph.cuh

```diff
@@ -0,0 +1,366 @@
+    }
+  }
+
+  //! \brief Constructs a `graph` object from a native CUDA graph handle.
+  //! \param __graph The native CUDA graph handle to construct the `graph` object from.
+  //! \throws None
+  //! \post `get() == __graph`
+  [[nodiscard]] _CCCL_HOST_API static _CCCL_CONSTEXPR_CXX20 auto from_native_handle(cudaGraph_t __graph) noexcept
+    -> graph
+  {
+    return graph{__graph};
+  }
+
+  //! \brief Adds a new root node to the graph.
+  //! \tparam _Node The type of the node to add.
+  //! \param __node The descriptor of the node to add to the graph.
+  //! \return A `graph_node_ref` representing the added node. The graph object owns the
+  //! new node.
+  //! \throws cuda::std::cuda_error if adding the node fails.
+  template <class _Node>
+  [[nodiscard]] _CCCL_TRIVIAL_HOST_API constexpr auto add(_Node __node) -> graph_node_ref
+  {
+    // assert that the node descriptor is not returning an owning graph_node object, which
+    // would cause this function to return a dangling reference.
+    static_assert(_CUDA_VSTD::_IsSame<decltype(__node.__add_to_graph(__graph_)), graph_node_ref>::value,
+                  "node descriptors must return a graph_node_ref");
+    return __node.__add_to_graph(__graph_);
+  }
+
+  //! \brief Adds a new node to the graph with specified dependencies.
+  //!
+  //! This function creates a new node in the graph and establishes dependencies
+  //! between the newly created node and the provided dependency nodes.
+  //!
+  //! \tparam _Node The type of the node to be added.
+  //! \tparam _Extent The extent of the span representing the dependencies.
+  //!
+  //! \param __node The descriptor of the node to be added to the graph.
+  //! \param __deps An array of `cudaGraphNode_t` handles representing the dependencies of
+  //! the new node. Each node in this span will become a dependency of the newly created
+  //! node.
+  //!
+  //! \return A `graph_node_ref` object representing the newly created node in the graph.
+  //! The graph object owns the new node.
+  //!
+  //! \throws cuda::std::cuda_error If the CUDA API call `cudaGraphAddDependencies` fails.
+  //!
+  //! \details
+  //! - The function first creates a new node in the graph using the provided `_Node` object.
+  //! - It initializes an array of "dependant" nodes, where all dependant nodes correspond
+  //!   to the newly created node.
+  //! - The function then uses the CUDA API `cudaGraphAddDependencies` to establish the
+  //!   dependencies between the newly created node and the nodes provided in the `__deps`
+  //!   span.
+  //! - If the number of dependencies is small, a stack-allocated buffer is used;
+  //!   otherwise, a dynamically allocated array is used to store the dependant nodes.
+  template <class _Node, size_t _Np>
+  _CCCL_HOST_API constexpr auto add(_Node __node, _CUDA_VSTD::array<cudaGraphNode_t, _Np> __deps) -> graph_node_ref
+  {
+    return add(_CCCL_MOVE(__node), _CUDA_VSTD::span{__deps});
```

Comment:
nitpick, the explicit construction of the span is unnecessary

Reply:
it _is_ necessary because otherwise this function would call itself recursively.